sklearn/neighbors/lof.py

```diff
@@ -0,0 +1,311 @@
+            dist = pairwise_distances(X, self._fit_X, 'euclidean',
+                                      n_jobs=self.get_n_jobs_, squared=True)
+        else:
+            dist = pairwise_distances(X, self._fit_X, self.effective_metric_,
+                                      n_jobs=self.get_n_jobs_,
+                                      **self.effective_metric_params_)
+
+        reach_dist_array = np.zeros((n_samples, self.n_neighbors))
+
+        for j in range(n_samples):
+            neighbors_number = -1
+            for i in neighbors_indices[j, :]:
+                neighbors_number += 1
+                if self.effective_metric_ == 'euclidean':
+                    reach_dist_array[j, neighbors_number] = np.max(
+                        [self._k_distance_value_fit_X_[i], sqrt(dist[j, i])])
+                else:
+                    reach_dist_array[j, neighbors_number] = np.max(
+                        [self._k_distance_value_fit_X_[i], dist[j, i]])
+
+        return 1. / np.mean(reach_dist_array, axis=1)
+
+    def _local_outlier_factor(self, X=None):
+        """Compute local outlier factor (LOF)
+
+        The local outlier factor (LOF) of a sample captures its
+        supposed `degree of abnormality'.
+        It is the average of the ratio of the local reachability density of
+        a sample and those of its k-nearest neighbors.
+
+        Parameters
+        ----------
+        X : array-like of shape (n_samples, n_features)
+            The points to compute the LOF w.r.t. training samples self._fit_X.
+            Note that samples X are not considered in the neighborhood of
+            self._fit_X for computing their local_reachability_density, but a
+            sample both in X and self._fit_X is counted twice (thus considered
+            in his own neighborhood).
+            If None, compute LOF of self._fit_X w.r.t. to itself. (In this case
+            samples are not counted twice, ie not considered in their own
+            neighbourhood).
+
+        Returns
+        -------
+        lof : array-like of shape (n_samples,)
+            The LOF of X. The lower, the more normal.
+        """
+        same_X = X is None
+        n_samples = self._fit_X.shape[0] if same_X else X.shape[0]
+        self._k_distance_value_fit_X_, self.neighbors_indices_fit_X_ = (
+            self._k_distance(None))
+        # Compute it in fit()?
+        # XXX May be optimized (if p is not None) by only computing it for
+        # X_sub_samples = neighbors of neighbors of X
+
+        neighbors_indices_X_ = (self.neighbors_indices_fit_X_ if same_X
+                                     else self._k_distance(X)[1])
+
+        # Compute the local_reachibility_density of samples X:
+        X_lrd = self._local_reachability_density(X)
+        lrd_ratios_array = np.zeros((n_samples, self.n_neighbors))
+
+        # Avoid re-computing X_lrd if same_X:
+        lrd = X_lrd if same_X else self._local_reachability_density(None)
+
```

Comment:
The variable `same_X` is not necessary and the use of the ternary operator complicate the understanding. 
Why don't use a unique bloc `if X is None : ... else : ...` ?